### ISSUE TITLE
Don't throw `unreplaced macro` when running without Vite

### ISSUE THREAD

**Issue by gustavopch** [pcattori/vite-env-only#19]
I am using `serverOnly$` to ensure Vite doesn't bundle a function that's supposed to be executed on the server. But I actually also need to run that function in a simple Node.js script that doesn't use Vite. When I run my script, I get this error:
```
Error: vite-env-only: unreplaced macro

Did you forget to add the 'vite-env-only' plugin to your Vite config?
```

So I propose that that error is only thrown when running via Vite.

Not sure if there's a better way to check, but possibly something like this:
```diff
 const maybe = <T>(_: T): T | undefined => {
+  if (
+    import.meta.env &&
+    'MODE' in import.meta.env &&
+    'BASE_URL' in import.meta.env &&
+    'PROD' in import.meta.env &&
+    'DEV' in import.meta.env &&
+    'SSR' in import.meta.env
+  ) {
     throw Error(
       [
         `${pkgName}: unreplaced macro`,
         "",
         `Did you forget to add the '${pkgName}' plugin to your Vite config?`,
         "👉 https://github.com/pcattori/vite-env-only?tab=readme-ov-file#installation",
       ].join("\n"),
     )
+  }
 }
```

**Comment by bradymwilliams** [comment]
```
diff --git a/node_modules/vite-env-only/dist/index.js b/node_modules/vite-env-only/dist/index.js
index 440090b..2fdec0f 100644
--- a/node_modules/vite-env-only/dist/index.js
+++ b/node_modules/vite-env-only/dist/index.js
@@ -280,6 +280,16 @@ var transform = (code, id, options) => {
 
 // src/macro.ts
 var maybe = (_) => {
+    const usingVite = import.meta.env &&
+    'MODE' in import.meta.env &&
+    'BASE_URL' in import.meta.env &&
+    'PROD' in import.meta.env &&
+    'DEV' in import.meta.env &&
+    'SSR' in import.meta.env
+
+    if (!usingVite) {
+    return _;
+  }
   throw Error(
     [
       `${name}: unreplaced macro`,

```

thanks, patched 2.2.0 with your fix

**Comment by pcattori** [comment]
Originally, I hesitated to implement this feature since it felt like the error provided a base layer of safety.
But when I tried to define what I meant by "safety" I realized that the error doesn't actually prevent the code from being included in the module graph.
It only replaces the runtime error, but not the code shipped to that environment.

An alternative I thought of was to have `unsafe_serverOnly$(...)` or `serverOnly$(..., { unsafe: true })` or something like that.
But that means opting in or out of safety where the macro is used, which is the wrong place; the whole point is that the same macro gets accessed sometimes from within Vite and sometimes outside of Vite.

Another option would be to have a `VITE_ENV_ONLY=server` or similar env var so that the error is only thrown when the macro is used outside of the expected env. But again, this doesn't prevent code from being included in the module graph.

So to summarize, I think you're right that a simple runtime check for Vite is the best approach. If you are using the Vite plugin, things continue to work the same way. If you aren't using the Vite plugin, there's no bundling going on so there's no module graph artifact to try to protect in the first place.

**Comment by juliuxu** [comment]
@pcattori Hey, any chance the original changes in the PR you made can be merged? I see you proposed a smarter solution, but perhaps this check is good enough to merge?

My team has the same problem, we are including some share code in our unit tests which triggers this error. We have been mitigating this by using `patch-package` as proposed. It works fine enough, but I'd really want to get it removed and have the fix/solution upstream instead.

**Comment by jrestall** [comment]
We are also hitting this issue, we have some code shared between our expressjs remix server and the remix bundle itself. The expressjs code is compiled with esbuild and therefore throws when using the same code, since there is no Vite plugin to remove the $serverOnly macros.

A solution that ensures $serverOnly doesn't throw when built outside of Vite such as with esbuild would be great.

**Comment by gustavopch** [comment]
@jrestall Besides the patch I provided in the first message, you can also work around by moving as much of your Express server code to the bundle as possible as described here: https://github.com/remix-run/remix/discussions/9790#discussioncomment-10625633

**Comment by jrestall** [comment]
Really appreciate the tip @gustavopch, thank you! Moving all our expressjs server code to entry.server would be a very large change across our apps but it's good to know there's another option. We'll go with the patch for now and hope the behavior is fixed.

**Comment by zwhitchcox** [comment]
Is there a temporary workaround? Or do we have to build from source with the patch? I don't see why we would care about people not using vite. The plugin is called `vite-env-only`.

**Comment by gustavopch** [comment]
@zwhitchcox You don't need to build from source to patch. You can use https://npm.im/package/patch-package or something similar.

**Comment by zwhitchcox** [comment]
@gustavopch Thanks!

**Comment by SylRob** [comment]
hitting this issue with playwright running e2e with dev server.
The patch indeed fox the issue but it would be great to have an official release for that

**Comment by Tony-Truand** [comment]
Hello,

I do own several websites using react-router + vite and thus your package is unavoidable.
But I got stuck with this, when i share code between website (using vite) and command line scripts.
Any chance to have this PR merged, please?

Merry christmas

**Comment by pcattori** [comment]
Apologies for not being more responsive to this thread earlier as I was waiting for more use-cases / issues to come in for this to see what a good design would be, but should have communicated that better.

The main hesitation I have is the following scenario:

- You install `vite-env-only`
- You import and use `serverOnly$` in some app code and some scripts
- You forgot to actually add the plugin in `vite.config.ts`
- You deploy your app

In that case, I want to make absolutely sure that server code does not get leaked.

Additionally, there's no guarantee that checking for stuff in `import.meta.env` means we are definitely running within a Vite environment; none of those env vars have `VITE` in them, so false positives could happen. Feels wrong to hard-code this into `vite-env-only` since the repercussions aren't just API ergonomics, but could have security implications.

So I agree a simple solution would be to just say that forgetting to add the plugin is user error since we document it, but saying "Aw too bad you held it wrong and leaked server code" when things go wrong feels like blaming the users for bad API design.

I think I prefer a design where the user explicitly opts in rather than implicitly trusting on brittle Vite detection. I'm open to suggestions, but so far still leaning towards `VITE_ENV_ONLY_MODE=unsafe` or something that can be set in the entrypoint of your scripts.

**Comment by pcattori** [comment]
Good point, I think you are right that those are the two main alternatives to consider.

I think it will be best to have some sort of environment declaration for the whole process (https://github.com/pcattori/vite-env-only/issues/19#issuecomment-2120712028), so scripts can do `VITE_ENV_ONLY=server` to allow `serverOnly$` to resolve without errors.

Maybe we need a better name or an `_UNSAFE` suffix, but yes I think env var is the way to go. Previously, I said that this could be brittle since if you accidentally import your script with `VITE_ENV_ONLY=server` into your client graph, then that would be bad. But Vite already handles env vars in the client with great care to avoid this case, plus the plugin could also look for `VITE_ENV_ONLY` in the source and fail if found

**Comment by gustavopch** [comment]
I agree and I think `VITE_ENV_ONLY=server` may actually be a good name. It's not necessarily that it's "unsafe", you're just telling it's running on the server, so naturally it should be allowed to see server-only things.

This env var will only exist because Vite doesn't offer a built-in `VITE_ENV=server|client` that would let us know the environment where the bundle is meant to run. If Vite ever introduces something like this in the future, `vite-env-only` may not need an env var of its own anymore.

**Comment by pcattori** [comment]
Hey everyone!

I've been working on updating `vite-env-only` to be compatible with [Vite's Environment API](https://vite.dev/guide/api-environment) and I want to make sure that next version also addresses this issue.

If you've commented before, could you please **share concrete example of the issue** you are facing?

To [reiterate what I mentioned earlier](https://github.com/pcattori/vite-env-only/issues/19#issuecomment-3829150247), I totally see how having the macro be a passthrough identity function outside of Vite would work, but it leaves the door open for server code to be accidentally leaked.

If anything, I would love for the macro to be imported via a [Vite virtual module](https://vite.dev/guide/api-plugin#importing-a-virtual-file) to ensure a build failure when server would have been included in the client module graph. That's a much stronger guarantee than hoping the runtime error is thrown for anyone who forgot to add the Vite plugin to their config.
